File: Dynamic Programming/minScoreTriangulationofPolygon1039.py
class Solution(object):
    def minScoreTriangulation(self, A):
        """
        :type A: List[int]
        :rtype: int
        """
        # top down DP
        memo = {}
        def dfs(i, j):
            if j - i < 2:
                return 0
            if (i, j) in memo:
                return memo[(i, j)]
            # memo[(i, j)] = min([dfs(i, k) + dfs(k, j) + A[i] * A[j] * A[k] for k in range(i + 1, j)] or [0])
            min_ = float('inf')
            for k in range(i + 1, j):
                min_ = min(dfs(i, k) + dfs(k, j) + A[i] * A[j] * A[k], min_)
            memo[(i, j)] = min_
            return memo[(i, j)]
        
        return dfs(0, len(A) - 1)

File: Dynamic Programming/minScoreTriangulationofPolygon1039.java
class Solution {
    public int minScoreTriangulation(int[] A) {
        // bottom up DP, dp[i][j] means the minimum score to triangulate A[i] ~ A[j]
        // Time O(n^3), Space O(n^2)
        // 0 <= i < k < j < n
        int n = A.length;
        int[][] dp = new int[n][n];
        for (int d = 2; d < n; d++) {
            for (int i = 0; i + d < n; i++) {
                int j = i + d;
                dp[i][j] = Integer.MAX_VALUE;
                for (int k = i + 1; k < j; k++) {
                    dp[i][j] = Math.min(dp[i][j], dp[i][k] + dp[k][j] + A[i] * A[j] * A[k]);
                }
            }
        }
        return dp[0][n - 1];
    }
}